call-home/src/bin/callhome/collector/report_models.rs

```diff
@@ -116,6 +206,232 @@ impl Nexus {
+    /// a new report_models::StorageNodes object by using the data provided.
+    pub(crate) fn new(
+        replicas: Vec<openapi::models::Replica>,
+        pools: Vec<openapi::models::Pool>,
+        node_disks: HashMap<String, Vec<BlockDevice>>,
+    ) -> Self {
+        let mut storage_nodes = Self::default();
+
+        // HashMap to keep track of counted node disks.
+        let mut counted_node_disks = HashMap::new();
+
+        // Iterate over each pool specification.
+        for pool_spec in pools.iter().filter_map(|pool| pool.spec.clone()) {
+            // Get or insert the node entry in storage_nodes.nodes for the pool's node
+            let node_entry = storage_nodes
+                .nodes
+                .entry(pool_spec.node.clone())
+                .or_insert(NodeInfo::default());
+
+            // Create a new PoolInfo instance with pool id and replicas, and push it to the node's
+            // pools
+            let pool_info = PoolInfo::new(pool_spec.id, replicas.clone());
+            node_entry.pools.push(pool_info);
+
+            // Retrieve the disks associated with the current pool's node.
+            if let Some(disks) = node_disks.get(&pool_spec.node) {
+                for pool_disk in &pool_spec.disks {
+                    // Retrieve the block device associated with the pool disk.
+                    if let Some(bdev) = get_bdev(disks, pool_disk) {
+                        // Determine the parent block device, if partition exists.
+                        let parent_bdev_name = bdev.partition.map(|partition| partition.name);
+
+                        // Count the disk if not already counted for the node.
+                        if counted_node_disks
+                            .entry(pool_spec.node.clone())
+                            .or_insert(HashSet::new())
+                            .insert(parent_bdev_name.unwrap_or(bdev.devname))
+                        {
+                            node_entry.spdk_managed_disks_count += 1;
+                        }
+                    }
+                }
+            }
+        }
+
+        storage_nodes
+    }
+}
+
+/// NodeInfo contains PoolInfo (replicas capacity and replicas count) for each pool in a node and
+/// spdk managed disks count for the node.
+#[derive(Serialize, Deserialize, Debug, Default)]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct NodeInfo {
+    pools: Vec<PoolInfo>,
+    spdk_managed_disks_count: u64,
+}
+
+/// PoolInfo contains total replicas capacity and replicas count in a pool.
+#[derive(Serialize, Deserialize, Debug, Default)]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct PoolInfo {
+    replicas: u64,
+    replicas_capacity_in_bytes: u64,
+}
+
+impl PoolInfo {
+    /// Receives pool_id and a vector of openapi::models::Replica and returns a new
+    /// report_models::PoolInfo object by using the data provided.
+    pub(crate) fn new(pool_id: String, replicas: Vec<openapi::models::Replica>) -> Self {
```

Comment:
Pass by reference instead, and avoid cloning

```suggestion
    pub(crate) fn new(pool_id: String, replicas: &Vec<openapi::models::Replica>) -> Self {
```

Reply:
Fixed